### PR TITLE
Stochastic Gradient Descent (SGD) Optimization

### PR DESCRIPTION
**Pull request by r-chong**
Key Changes:
- Add optimizer class
- Implement stochastic gradient descent (extends optimizer)
- Complete 1.5

### CODE REVIEW

**Comment by r-chong** [comment]
Nice
image

**Comment by r-chong** [comment]
Found optimal training for all but spiral

```ts
  const data1 = datasets["Simple"](PTS) as Graph;
  const data2 = datasets["Diag"](PTS) as Graph;
  const data3 = datasets["Split"](PTS) as Graph;
  const data4 = datasets["Xor"](PTS) as Graph;
  const data5 = datasets["Circle"](PTS) as Graph;
  const data6 = datasets["Spiral"](PTS) as Graph;

  console.log("=== Simple [4] ===");
  new ScalarTrain([4]).train(data1, RATE);
  console.log("=== Diag [4] ===");
  new ScalarTrain([4]).train(data2, RATE);
  console.log("=== Split [8] ===");
  new ScalarTrain([8]).train(data3, RATE);
  console.log("=== Xor [8] ===");
  new ScalarTrain([8]).train(data4, 0.5, 1000);
  console.log("=== Circle [8, 8] ===");
  new ScalarTrain([8, 8]).train(data5, 0.5, 1000);
```

**Comment by r-chong** [comment]
```ts
=== Simple ===
Epoch  500  loss  0.03617757487190162  correct  50/50

=== Diag ===
Epoch  500  loss  0.054961770497248166  correct  49/50

=== Split ===
Epoch  500  loss  0.08101704169346319  correct  50/50

=== Xor ===
Epoch  500  loss  0.36880885965233634  correct  43/50

=== Circle ===
Epoch  1000  loss  0.07049113245985564  correct  49/50

=== Spiral ===
Epoch  1000  loss  0.6708060423131487  correct  30/50
```

See https://github.com/r-chong/TSTorch/pull/12 for spiral optimizations

File: packages/tstorch/src/module.ts [hunk 1]
```diff
@@ -95,6 +96,13 @@ export class Parameter<T=Tensor> extends BaseParameter {
         }
     }
 
+    get grad() {
+        if (this.value instanceof Scalar) {
+            return this.value.derivative ?? 0;
```
Comment (MankyDanky) [inline packages/tstorch/src/module.ts:106]: You might want to consider initializing the derivative as `null` since it's a Scalar class, not sure 100%

File: packages/tstorch/src/optimizer.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                continue;
+            }
+
+        // Check for derivative (Scalar-like objects)
+        if (p.value instanceof Scalar) {
+            const grad = p.value.derivative ?? 0;
+            p.value.data -= this.lr * grad;
+        }
+    }
+}
```
Comment (MankyDanky) [inline packages/tstorch/src/optimizer.ts]: Odd indentation here

File: packages/tstorch/src/scalar.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ let _varCount = 0;
  * Behaves like a regular number but records operations for autodiff.
  */
 export class Scalar {
-    readonly data: number;
+    data: number;
```
Comment (MankyDanky) [inline packages/tstorch/src/scalar.ts:28]: Why did we make this not readonly anymore? I don't think it should be mutable right as this messes with the computation graph

Comment (r-chong) [reply packages/tstorch/src/scalar.ts:28]: In `optimizer.ts`, I did

```ts
if (p.value instanceof Scalar) {
            const grad = p.value.derivative ?? 0;
            p.value.data -= this.lr * grad;
        }
```

However, if the data is read-only, this does not work. I tried doing
```ts
p.update(new Scalar(p.value.data - lr * grad))
```
However this breaks autograph. It creates a new node, but the old node is still in the computation graph.

Comment (r-chong) [reply packages/tstorch/src/scalar.ts:28]: This is my understanding:

- During the forward pass, we run computations and save them in the computation graph
- During the backward pass, we partially differentiate said computations and save in .grad
- After both are completed, we update, at which point it is okay to mutate data

> If we create a new Scalar during the optimizer step (instead of mutating the existing one), we replace the parameter object. The computation graph still holds references to the original Scalar used in the forward pass, so during backward gradients accumulate into the old object, while the optimizer updates the new one. Because the graph and the model now refer to different objects, gradients never affect the actual weights and training breaks. 

(ChatGPT)

File: packages/tstorch/src/optimizer.ts (hunk 2 above)
Comment (r-chong) [reply packages/tstorch/src/optimizer.ts]: Fixed

File: packages/tstorch/src/module.ts (hunk 1 above)
Comment (r-chong) [reply packages/tstorch/src/module.ts:106]: For backpropagation, we sum the contributions (derivatives) so they should be a number (0)
